File: 백준/Silver/14501. 퇴사/퇴사.js
const fs = require('fs');
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
// const input = fs.readFileSync('./예제.txt').toString().split('\n');
let n = Number(input.shift())
let arr = input.map(v=>v.split(' ').map(Number))
const dp = Array(n).fill(0)

for(let i = 0 ; i < n ; i++){
    const [day, pay] = arr[i]

    if(i + day > n){
        continue
    }

    dp[i] += pay

    for(let j = i + day ; j < n; j++){
        dp[j] = Math.max(dp[j], dp[i])
    }
}

console.log(Math.max(...dp))

File: 백준/Silver/14501. 퇴사/퇴사.py
from sys import stdin
input = stdin.readline

n = int(input())
arr = [[0, 0]]
for _ in range(n):
    arr.append(list(map(int, input().split())))

dp = [0 for _ in range(n+1)]

for i in range(n+1): # 상담에 필요한 기간이 퇴사날을 벗어나는 경우 제거
    if i + arr[i][0] > n+1:
        arr[i][1] = 0

for i in range(n): # dp를 이용해 각 날짜의 최댓값을 구하기
    for j in range(i+arr[i][0], n+1):
        if dp[j] < dp[i] + arr[j][1]:
            dp[j] = dp[i] + arr[j][1]

print(max(dp)) # dp의 최댓값 -> 최대비용